aten/src/ATen/FunctionalTensorWrapper.cpp

```diff
@@ -322,6 +322,57 @@ void sync(const c10::List<c10::optional<Tensor>> t_list) {
   }
 }
 
+bool isFunctionalTensor(const at::Tensor& tensor) {
+  return tensor.unsafeGetTensorImpl()->key_set().has(c10::DispatchKey::Functionalize);
+}
+
+TORCH_API bool isFunctionalTensor(const c10::optional<Tensor>& t) {
+  if (t.has_value()) {
+    return isFunctionalTensor(*t);
+  } else {
+    return false;
+  }
+}
+
+TORCH_API bool isFunctionalTensor(const c10::List<Tensor>& t_list) {
+  if (t_list.size() == 0) return false;
+  bool any_functional = isFunctionalTensor(t_list[0]);
+  for (const auto i : c10::irange(1, t_list.size())) {
+    auto curr_functional = isFunctionalTensor(t_list[i]);
+    TORCH_INTERNAL_ASSERT(
+         curr_functional == any_functional,
+        "Functionalization encountered a list of tensors where some are functional",
+        "and some are not, which is unsupported.");
```

Comment:
Unsupported here just means "we didn't bother implement it" as opposed to "fundamentally this is impossible to support" right?

Reply:
yep that's right (I'll make the comment a bit clearer).